services/datamanager/builtin/file_deletion.go

```diff
@@ -0,0 +1,90 @@
+package builtin
+
+import (
+	"errors"
+	"io/fs"
+	"path/filepath"
+	"strings"
+
+	"go.viam.com/rdk/logging"
+
+	"github.com/ricochet2200/go-disk-usage/du"
+	"go.viam.com/rdk/services/datamanager/datasync"
+)
+
+const (
+	fileDeletionThreshold    = .95
+	captureDirRatioThreshold = .5
+	n                        = 4
+)
+
+var errAtSizeThreshold = errors.New("capture dir is at correct size")
+
+func checkFileSystemStats(captureDirPath string, logger logging.Logger) (bool, error) {
+	usage := du.NewDiskUsage(captureDirPath)
+	usedSpace := usage.Usage()
+	if usedSpace < fileDeletionThreshold {
+		logger.Debug("Should exit thread, disk not full enough ignoring for now")
+		// return false, nil
+	}
+	//walk the dir to get capture stats
+	return checkCaptureDirSize(captureDirPath, float64(usage.Size()), logger)
+}
+
+func checkCaptureDirSize(captureDirPath string, fsSize float64, logger logging.Logger) (bool, error) {
+	var dirSize int64 = 0
+	logger.Warn("checking size")
+
+	readSize := func(path string, d fs.DirEntry, err error) error {
+		if !d.IsDir() {
+			fileInfo, err := d.Info()
+			if err != nil {
+				return err
+			}
+			// logger.Warn(fileInfo.Name())
+			dirSize += fileInfo.Size()
+			if float64(dirSize)/fsSize < captureDirRatioThreshold {
+				logger.Warnw("At threshold to delete, going to delete", "size", float64(dirSize)/fsSize, "threshold", captureDirRatioThreshold)
+				return errAtSizeThreshold
+			} //else {
+			// 	logger.Warnw("Not at threshold", "size", float64(dirSize)/fsSize, "threshold", captureDirRatioThreshold)
+			// }
+		}
+		return nil
+	}
+
+	err := filepath.WalkDir(captureDirPath, readSize)
+	if err != nil && !errors.Is(err, errAtSizeThreshold) {
+		return false, err
+	}
+	return err != nil && errors.Is(err, errAtSizeThreshold), nil
+}
+
+func deleteFiles(syncer datasync.Manager, captureDirPath string, logger logging.Logger) error {
+	index := 0
+	delete := func(path string, d fs.DirEntry, err error) error {
+		if !d.IsDir() {
+			fileInfo, err := d.Info()
+			if err != nil {
+				return err
+			}
+			logger.Debug(fileInfo.Name())
+			isFileInProgress := strings.Contains(fileInfo.Name(), ".prog")
+			// if at nth file, the file is not currenlty being written, the syncer isnt nil and isnt uploading the file or there is no syncer
+			if index%n == 0 && !isFileInProgress && ((syncer != nil && syncer.MarkInProgress(path)) || syncer == nil) {
```

Comment:
I think calling MarkInProgress is problematic here. Say the syncer has already marked in the file in progress for deletion. This call will release the lock on that file, is that what we want to do? I would think if the another process has acquired a lock on the specified, we would respect it and try a different file.

Reply:
Talked about this offline but MarkInProgress returns false if a file is already marked as in progress and does not change. This if statement will exit if MarkInProgress returns false.

Reply:
That's my bad, instead of returning false, I thought it was changing the in progress value to false. :)